File: awesome/.config/awesome/util/screenshot.lua
-----------------------------------------------------------------------
-- Fern Zapata
-- https://github.com/fernzi/dotfiles
-- Awesome Window Manager - Screenshot/cast Utilities
-----------------------------------------------------------------------

local awful = require('awful')
local beautiful = require('beautiful')
local gears = require('gears')
local naughty = require('naughty')

local settings = require('settings')
local path = require('util.path')
local audio = require('util.audio')

local dpi = beautiful.xresources.apply_dpi

local screenshot = {}

-----------------------------------------------------------------------

local screenshot_path = settings.screenshot_path
                     or path.pictures .. 'Screenshots/'
local screencast_path = settings.screencast_path
                     or path.videos .. 'Screencasts/'

local function _run(command, file)
  gears.filesystem.make_parent_directories(file)
  return awful.spawn.easy_async(command, function()
    if not gears.filesystem.file_readable(file) then return end
    audio.play('screen-capture')
    naughty.notify {
      title = 'Screen captured!',
      text = file:gsub(os.getenv('HOME'), '~', 1),
      icon = file,
      icon_size = dpi(64),
      actions = {
        Open = function()
          awful.spawn{settings.file_manager, file}
        end
      }
    }
  end)
end

function screenshot.full()
  local file = screenshot_path .. os.date('%F_%T.png')
  local command = {'maim', file}
  _run(command, file)
end

function screenshot.window()
  if not client.focus then return end
  local file = screenshot_path .. os.date('%F_%T.png')
  local command = {'maim', '-i', tostring(client.focus.window), file}
  _run(command, file)
end

function screenshot.area()
  local file = screenshot_path .. os.date('%F_%T.png')
  local command = {'maim', '-s', file}
  _run(command, file)
end

local recording = 0

function screenshot.record()
  local file = screencast_path .. os.date('%F_%T.mp4')
  local scn = awful.screen.focused()
  local command = {
    'ffmpeg',
    '-vaapi_device', '/dev/dri/renderD128',
    '-f', 'x11grab',
    '-s', ('%dx%d'):format(scn.geometry.width, scn.geometry.height),
    '-r', '15',
    '-i', ('%s+%d,%d'):format(os.getenv('DISPLAY'), scn.geometry.x, scn.geometry.y),
    '-vf', 'hwupload,scale_vaapi=format=nv12',
    '-c:v', 'h264_vaapi',
    '-qp', '18',
    '-y',
    file,
  }
  if recording == 0 then
    recording = _run(command, file)
  elseif awesome.kill(recording, awesome.unix_signal.SIGTERM) then
    recording = 0
  end
end

-----------------------------------------------------------------------

setmetatable(screenshot, {
  __call = function(self, ...)
    return self.full(...)
  end,
})

return screenshot

File: qtile/.config/qtile/util/screenshot.py
# Fern Zapata
# https://github.com/fernzi/dotfiles
# Qtile Window Manager - Screenshot/cast utilities

import asyncio as aio
import time
from pathlib import Path

from libqtile import qtile
from gi.repository import Notify

import settings
from util import audio


def _screenshot_path() -> Path:
  try:
    path = Path(settings.screenshot_path)
  except AttributeError:
    path = Path('~/Pictures').expanduser()
  return path / 'Screenshots' / time.strftime('%F_%T.png')


async def _run(command: list, save: Path) -> bool:
  save.parent.mkdir(parents=True, exist_ok=True)
  proc = await aio.create_subprocess_exec(*command)
  await proc.wait()
  if save.exists():
    audio.play('screen-capture')
    if Notify.is_initted():
      Notify.Notification.new(
        'Screen captured!',
        str(save),
        str(save),
      ).show()
    return True
  return False


async def full() -> bool:
  save = _screenshot_path()
  command = ['maim', save]
  return await _run(command, save)


async def area() -> bool:
  save = _screenshot_path()
  command = ['maim', '-s', save]
  return await _run(command, save)


async def window() -> bool:
  try:
    save = _screenshot_path()
    window_id = str(qtile.current_window.info()['id'])
    command = ['maim', '-i', window_id, save]
    return await _run(command, save)
  except AttributeError:
    return False
